text/0000-node-api-v2.md

```diff
@@ -0,0 +1,1131 @@
+              }
+            ]
+          },
+          "offset": "0ec14d3875ad5a366418256fe65bad2a4d4ff1914e1b9488db72dd355138ca3a"
+        },
+        true
+    ],
+    "id": 1
+}
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "result": {
+    "Ok": null
+  }
+}
+```
+
+## Errors
+
+There is two kind of errors that can appear when making a call on the API v2: basic JSON-RPC errors and API result errors
+
+### JSON-RPC Errors
+
+These errors are often due to a miscontructed JSON body.
+For example when there is not enough parameters:
+
+```JSON
+{
+  "error": {
+    "code": -32602,
+    "message": "WrongNumberOfArgs. Expected 4. Actual 5"
+  },
+  "id": 1,
+  "jsonrpc": "2.0"
+}
+```
+
+or invalid type of data entered:
+
+```JSON
+{
+  "error": {
+    "code": -32602,
+    "message": "InvalidArgStructure \"start_height\" at position 1."
+  },
+  "id": 1,
+  "jsonrpc": "2.0"
+}
+```
+
+or method not found:
+
+```JSON
+{
+  "error": {
+    "code": -32601,
+    "message": "Method not found"
+  },
+  "id": 2,
+  "jsonrpc": "2.0"
+}
+```
+
+this list of errors is not exhaustive, for more information about the possible error objects see the [JSON RPC 2.0 specifications](https://www.jsonrpc.org/specification#error_object).
+
+
+### API Result Errors
+
+These type of errors are due to an API error during the query. This type of error is wrapped into the result. 
+The possible error objects are defined in the [`grin_api` crate](https://docs.rs/grin_api/2.0.0/grin_api/struct.Error.html)
+
+For example, trying to ban a peer that's already banned:
+
+```JSON
+{
+  "id": 2,
+  "jsonrpc": "2.0",
+  "result": {
+    "Err": {
+      "Internal": "ban peer error: PeerNotFound"
+    }
+  }
+}
+```
+
+or a block that doesn't exist:
+
+```JSON
+{
+  "id": 2,
+  "jsonrpc": "2.0",
+  "result": {
+    "Err": "NotFound"
+  }
+}
+```
+
+## Authentication
+
+Like the v1 API, the v2 API uses basic auth with the same secret. This token is usually in `grin/main/.api_secret`.
```

Comment:
Perhaps this model is too simple, we clearly have more than one API here, at least 2 - data consumer and node management (perhaps also data producer - `push_tx`), which may require different tokens or more granular auth model.

Reply:
Clearly we could separate them in 2 or 3 difference endpoints. Do we want to add such complexity though. For such use case someone can just build a middleware on top of it to filter the requests?

Reply:
Logically we have 2 APIs here, let's call them Blockchain API and Node API. The former is not node-specific (at least eventually), the later is. There are 2 different use cases and sets of clients. Security-wise exposing node api to internet is very risky. It's not that easy to restrict access to different JSON-RPC calls as for REST API (simple URL filtering is enough).
Token per endpoint model doesn't seem too hard to me, 1-2 lines of code to lookup a config key eg `api_token_{endpoint_name}`

Reply:
I see what you mean. However I'm still not sure it's worth the additional complexity here. It's also possible to do filtering with JSON RPC method. Not against but not convinced :)

Reply:
I agree with this concern, it might be worthwhile to split them. A wallet only needs very few of these endpoints to fully function, so anyone exposing their node publically would prefer to only have those few methods available on the outside.